drivers: stepper: add stepper driver for allegro a4979

verenabt opened the pull request:
This PR adds a stepper driver for [allegro a4979](https://www.allegromicro.com/-/media/files/datasheets/[a4979-datasheet.pdf](https://www.allegromicro.com/-/media/files/datasheets/a4979-datasheet.pdf)) microstepping driver.

It has manly been developed analogous to [drv8429](https://github.com/zephyrproject-rtos/zephyr/tree/main/drivers/stepper/ti) with adaptions where they are needed. It was tested with the [generic stepper sample](https://github.com/zephyrproject-rtos/zephyr/pull/85757).

faxe1008 reviewed (commented):
Nice work, some things around formatting and minor nits 👍

jilaypandya reviewed (commented):
Thanks @verenascst for the contribution :-) Try running 
 `python scripts/ci/check_compliance.py` after you have committed your changes locally and it should report all the whitespacing & other compliance errors :-)

jilaypandya commented on drivers/stepper/allegro/Kconfig:
```diff
@@ -0,0 +1,17 @@
+# SPDX-FileCopyrightText: Copyright (c) 2025 Carl Zeiss Meditec AG
+# SPDX-License-Identifier: Apache-2.0
+
+config STEPPER_ALLEGRO
```
```suggestion
menuconfig STEPPER_ALLEGRO
```

jilaypandya commented on drivers/stepper/allegro/Kconfig.a4979:
```diff
@@ -0,0 +1,11 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config A4979_STEPPER
+	bool "Activate driver for A4979 stepper motor driver"
```
```suggestion
	bool "Activate allegro A4979 stepper driver"
```

jilaypandya commented on drivers/stepper/allegro/Kconfig.a4979:
```diff
@@ -0,0 +1,11 @@
+	default y
+	depends on DT_HAS_ALLEGRO_A4979_ENABLED
+	select STEP_DIR_STEPPER
+	select STEPPER_STEP_DIR_GENERATE_ISR_SAFE_EVENTS
```
User should be able to select this :)
```suggestion
```

jilaypandya commented on drivers/stepper/allegro/a4979.c:
```diff
@@ -0,0 +1,291 @@
+#include "../step_dir/step_dir_stepper_common.h"
+
+#include <zephyr/logging/log.h>
+
```
```suggestion

```

jilaypandya commented on drivers/stepper/allegro/a4979.c:
```diff
@@ -0,0 +1,291 @@
+ * 
+ * This structure contains mutable data used by a A4979 stepper driver.
+ */
+struct a4979_data {
```
Same as what @faxe1008 mentioned above :)

jilaypandya commented on dts/bindings/stepper/allegro/allegro,a4979.yml:
```diff
@@ -0,0 +1,46 @@
+description: |
```
The spacing needs to be adjusted over here a bit :)

[tmc2209](https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/stepper/adi/adi%2Ctmc2209.yaml) is a good example.
https://docs.zephyrproject.org/latest/build/dts/api/bindings/stepper/adi/adi%2Ctmc2209.html

jbehrensnx commented:
Hi, could you try running the drv8424/api testsuite on your driver? It isnt yet generalized for more stepper drivers beside the drv824, but your driver is similar that you might still be able to run it.

faxe1008 commented on drivers/stepper/allegro/a4979.c:
```diff
@@ -0,0 +1,277 @@
+			dev->name);
+		return -ENOTSUP;
+	}
+
```
Disabling of the timing source was fine in case of calling this with `false` what I meant with my comment is if there is a benefit to setting the step pin to `0` after disabling it.

jbehrensnx replied:
This is taken from the drv824 driver. The drv8424 only takes a step on a rising edge. Thus it is there to ensure that the step pin has the correct value when a new movement command is called. Before step-dir was separated, the step pin was in state high for half of the step time and thus the disable command could otherwise leave the step pin in high, which would lead to a missing step for the new movement command.
I am not sure if this can still happen with the current step-dir implementation, as it toggles the pin twice in the same interrupt. That said, it still ensures the correct pin value, but one could argue, that it better fits during enabling.

verenascst replied:
Just realized during testing that disabling of the timing source now was missing. 
The a4979 also only takes a step on a rising edge. The implementation does work without ensuring the step pin having the correct value before a new movement command is called. Still I think it would make sense to set the step pin to zero during the init function. What are your opinions?

jbehrensnx replied:
0 on init would make sense, through 0 on enable (before setting enable pin to 1) might also make sense. At least on the drv8424 the step pin value is irrelevant while disabled but becomes important once enabled. I admit, I dont know what happens on the driver when you enable with step pin on 1. That said, this depends on how one views things happening on the step pin while disabled. If this is classified as undefined behavior, then 0 on init is enough.

jbehrensnx commented:
I think there is a misunderstanding. I just wanted to ask you to run the drv8424/api testsuite on your driver by adding/modifying an overlay file, not create a new test suite.
Or are the test fragments in the last commit not meant to be included?

verenascst commented:
> I think there is a misunderstanding. I just wanted to ask you to run the drv8424/api testsuite on your driver by adding/modifying an overlay file, not create a new test suite. Or are the test fragments in the last commit not meant to be included?

Yes, I think I misunderstood your request. 😅 
Still, the drv8424/api testsuite seems to be specifically named for drv8424 and I would keep an overlay for the a4979 rather downstream. Maybe it would be helpful to have a generic test for gpio-steppers. What would you suggest @jbehrensnx?

jbehrensnx commented on drivers/stepper/allegro/a4979.c:
```diff
@@ -0,0 +1,281 @@
+		break;
+	default:
+		LOG_ERR("Unsupported micro step resolution %d", micro_step_res);
+		return -EINVAL;
```
This should return `-ENOTSUP`. This is an error in the drv8424 driver and test suite that has not yet been fixed.

jbehrensnx commented:
> > I think there is a misunderstanding. I just wanted to ask you to run the drv8424/api testsuite on your driver by adding/modifying an overlay file, not create a new test suite. Or are the test fragments in the last commit not meant to be included?
> 
> Yes, I think I misunderstood your request. 😅 Still, the drv8424/api testsuite seems to be specifically named for drv8424 and I would keep an overlay for the a4979 rather downstream. Maybe it would be helpful to have a generic test for gpio-steppers. What would you suggest @jbehrensnx?

[PR](https://github.com/zephyrproject-rtos/zephyr/pull/85658) is already up, but there are some RFCs/PRs concerning some upcoming changes to the stepper api, so it on hold until they are finished.

jilaypandya commented on tests/drivers/stepper/a4979/api/src/main.c:
```diff
@@ -0,0 +1,476 @@
+	int32_t steps = 100;
+	bool moving = true;
+
+	(void)stepper_enable(fixture->dev, true);
```
Are these tests taken over completely from drv8424?

jilaypandya commented on drivers/stepper/allegro/a4979.c:
```diff
@@ -0,0 +1,281 @@
+	.set_event_callback = step_dir_stepper_common_set_event_callback
+};
+
+#define A4979_DEVICE(inst)                                              \
```
you need to correct this spacing manually

jilaypandya commented on dts/bindings/stepper/allegro/allegro,a4979.yml:
```diff
@@ -0,0 +1,49 @@
+
+  Example:
+  a4979: a4979 {
+    status = "okay";
```
same  for the other fields :)
```suggestion
      status = "okay";
```

jilaypandya commented on tests/drivers/build_all/stepper/gpio.dtsi:
```diff
@@ -37,3 +37,16 @@ drv8424: drv8424 {
 	counter = <&counter0>;
 };
+
+test_a4979: a4979_stepper {
```
building upon this https://github.com/zephyrproject-rtos/zephyr/pull/86685/files :) Depending on which PR gets merged, we might run into a merge conflict

Woud be great if this could be put on the top of the list.

```suggestion
allegro_a4979: allegro_a4979 {
```

jilaypandya reviewed (commented):
LGTM, just some few minor Nits and we are good to go :) 

On c files `clang-format -i path/to/c_file` should help in fixing the trailing whitespace issue. In the other files i would have done manually. 

Probably squash all the commits and format the message with line length of < 75, then that issue would be solved as well :)

jilaypandya reviewed (dismissed):
Thanks :-)

faxe1008 reviewed (commented):
If you can, adding a `stop` api impl. would be lovely.

jilaypandya commented:
No merge commits :). You need to rebase branch on main :)

fabiobaltieri commented on drivers/stepper/allegro/a4979.c:
```diff
@@ -0,0 +1,280 @@
+	bool has_enable_pin = config->en_pin.port != NULL;
+
+	/* Check availability of enable pin, as it might be hardwired. */
+	if (has_enable_pin) {
```
invert the condition, return the error in the branch and keep the normal exection unindented

fabiobaltieri commented on drivers/stepper/allegro/a4979.c:
```diff
@@ -0,0 +1,280 @@
+			return ret;
+		}
+	} else {
+		LOG_ERR("%s: Failed to enable/disable device, enable pin is not available. "
+			"The device is always on.",
+			dev->name);
+		return -ENOTSUP;
```
These messages take flash all the time and normally only have a chance to fire duing development, I'd recommend either keeping them as short as possible ("enable pin undefined") or changing it to `LOG_DBG`

verenascst commented:
> If you can, adding a `stop` api impl. would be lovely.

I won't be able to add the `stop` api impl. in this PR, but I appreciate the suggestion!

dipakgmx commented:
Nit: the commits need to be split since the changes are across 3 different directories

verenascst commented:
> Nit: the commits need to be split since the changes are across 3 different directories

So you mean like that? :)
    1. drivers: stepper: Add stepper driver for allegro a4979
    2. dts: bindings: stepper: Add binding for allegro a4979
    3. tests: drivers: Add test case for allegro a4979

Commit:
drivers: stepper: Add stepper driver for allegro a4979

Adding a stepper driver implementation for allegro a4979
microstepping programmable stepper motor driver.
The implemenation was tested using the drv8424/api testsuite.

Signed-off-by: Verena Schweinstetter <verena.schweinstetter@zeiss.com>

jilaypandya commented:
> > > Nit: the commits need to be split since the changes are across 3 different directories
> > 
> > 
> > So you mean like that? :) 1. drivers: stepper: Add stepper driver for allegro a4979 2. dts: bindings: stepper: Add binding for allegro a4979 3. tests: drivers: Add test case for allegro a4979
> 
> Yes

Commit 1 cannot work without Commit 2. Hence, they need to be squashed together. test commit can be kept seperate :)

Commit:
tests: drivers: Add test case for allegro a4979

Adding a test case for stepper driver implementation of
allegro a4979 microstepping programmable stepper motor driver.

Signed-off-by: Verena Schweinstetter <verena.schweinstetter@zeiss.com>

jilaypandya commented:
If you want you can add your driver to the stepper api tests as well
